middle_end/flambda/terms/let_expr.rec.ml

```diff
@@ -30,10 +30,49 @@ let pattern_match t ~f =
   A.pattern_match t.name_abstraction
     ~f:(fun bindable_let_bound body -> f bindable_let_bound ~body)
 
-let pattern_match_pair t1 t2 ~f =
-  A.pattern_match_pair t1.name_abstraction t2.name_abstraction
-    ~f:(fun bindable_let_bound body1 body2 ->
-      f bindable_let_bound ~body1 ~body2)
+module Pattern_match_pair_error = struct
+  type t = Mismatched_let_bindings
+
+  let to_string = function
+    | Mismatched_let_bindings -> "Mismatched let bindings"
+end
+
+let pattern_match_pair t1 t2 ~dynamic ~static =
+  A.pattern_match t1.name_abstraction
+    ~f:(fun bindable_let_bound1 body1 ->
+      A.pattern_match t2.name_abstraction
+        ~f:(fun bindable_let_bound2 body2 ->
+          let case =
+            match bindable_let_bound1, bindable_let_bound2 with
+            | Bindable_let_bound.Singleton _,
+              Bindable_let_bound.Singleton _ ->
+              `Dynamic
```

Comment:
Can you use a regular variant type here?  We've generally been steering clear of polymorphic variants (they make error messages worse, and can be slower to match on).

Reply:
Ended up rewriting using a local function for the dynamic cases.